File: niles.php
<?php

session_start();

//  This program builds and then sends a control string to the Niles GXR2 based on inputs received from controller.php

$server_ip = '10.100.0.1'; 
$server_port = 6001;  
$InputSelector = $_POST["InputSelector"]; 
$zone = $_SESSION["Zone"];
$navigation = $_SESSION["navigation"]; 

// Choose zone ID

if ($zone == 'Living Room') 
{ 
  $zoneid = "\x21";
}
elseif ($zone == 'Kitchen')
{
  $zoneid = "\x22";
}
elseif ($zone == 'Zone 3')
{
  $zoneid = "\x23"; 
}
elseif ($zone == 'Patio')
{
  $zoneid = "\x24"; 
}
elseif ($zone == 'M Bath')
{
  $zoneid = "\x25"; 
}
elseif ($zone == 'Basement')
{
  $zoneid = "\x26"; 
}
else 
{
print "No Zone Match <br>";
}

// Choose input ID

if ($InputSelector == 'Input1') 
{
  $inputid = "\x01";
}
elseif ($InputSelector == 'Input2')
{
  $inputid = "\x02";
}
elseif ($InputSelector == 'Input3')
{
  $inputid = "\x03"; 
}
elseif ($InputSelector == 'Input4')
{
  $inputid = "\x04"; 
}
elseif ($InputSelector == 'Input5')
{
  $inputid = "\x05"; 
}
elseif ($InputSelector == 'Input6')
{
  $inputid = "\x06"; 
}
elseif ($InputSelector == 'Vol+')
{
  $inputid = "\x0c"; 
}
elseif ($InputSelector == 'Vol-')
{
  $inputid = "\x0d";
}
elseif ($InputSelector == 'Prev')
{
  $inputid = "\x2b";
}
elseif ($InputSelector == 'Next')
{
  $inputid = "\x2c"; 
}
elseif ($InputSelector == 'Off')
{
  $inputid = "\x0a"; 
}
else {
print "No Input Match <br>";
}

// Build control string and store it in $message

$message = "\x00\x12\x00" . $zoneid . "\x00\x0b\x61\x06" . $inputid . "\x00\xff";

//  Open a UDP socket

if ($socket = socket_create(AF_INET, SOCK_DGRAM, SOL_UDP)) {
    socket_bind($socket, $_SESSION["LocalIP"], 6001);
  
//  Send the message
  
socket_sendto($socket, $message, strlen($message), 0, $server_ip, $server_port);
}
else {
  print("can't create socket\n");
}


// If the control string is for vol+ or vol-, send it five times because volume changes for individual messages are
// very small

if ($InputSelector == 'Vol*'){
    for ($x=1; $x<=5; $x++){
         socket_sendto($socket, $message, strlen($message), 0, $server_ip, $server_port);
    }
}

echo "<meta http-equiv = \"refresh\" content = \"1; URL=http://" . $_SESSION["LocalIP"] . "/controller.php\">";

File: niles.cpp
#include <iostream>
#include <string>
#include <cstring>
#include <sstream>
#include <iomanip>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <sys/types.h>

using namespace std;

int main()
{
    string result;
    string word[10];
    char zoneid = 0x00;
    char inputid = 0x00;
    int i, j, k ,l;
    char outbuffer[] = {0x00, 0x12, 0x00, zoneid, 0x00, 0x0b, 0x61, 0x06, inputid, 0x00, 0xff};
    char zones[] = {0x21, 0x22, 0x23, 0x24, 0x25, 0x26};
    char inputs[] = {0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06};
    string nums[] = {"0", "1", "2", "3", "4", "5", "6"};
    int zoneindex, inputindex;
    

    int sendfd = 0,connfd = 0;
  
    struct sockaddr_in serv_addr;

    sendfd = socket(AF_INET, SOCK_DGRAM, 0);
    serv_addr.sin_family = AF_INET;    
    serv_addr.sin_addr.s_addr = inet_addr("10.100.0.1");
    serv_addr.sin_port = htons(6001);    
    bind(sendfd, (struct sockaddr*)&serv_addr,sizeof(serv_addr));

    i=1;
    while (i==1)
    {
        cout << endl <<"Enter Command: " << endl;
        getline(cin, result);
        stringstream iss(result);
        j=0;
        while (iss >> word[j])
        {
            cout << word[j] <<endl;
            j++;
        }
        if((word[0] == "Zone") && (word[2] == "Input"))
        {
            for(k=0; k<=6; k++)
            {
                if (word[1] == nums[k])
                {
                    zoneindex = k-1;  // adjust zone index to start at zero
                }
                if (word[3] == nums[k])
                {
                    inputindex = k;
                }
            }
            printf("%d   %d\n", zoneindex, inputindex);
            outbuffer[3] = zones[zoneindex];
            outbuffer[8] = inputs[inputindex];
            sendto(sendfd,outbuffer,sizeof(outbuffer),0,(struct sockaddr *)&serv_addr,sizeof(serv_addr));
    
        }
        else
        {
            i=0;
        }
    }
    
    close(sendfd);    
    
    return (0);
}
